### PR TITLE
🏗✨ Dynamically generate custom closure compiler `runner.jar`

### LINKED ISSUE #22452: Dynamically generate custom closure compiler for AMP
Now that AMP's closure compiler is at the latest available version (#22446) modulo one awaited bug fix (https://github.com/google/closure-compiler/issues/3041), it's time to think about how to stay there as new releases appear.

**Here's my proposed solution:**
1. Wait for https://github.com/google/closure-compiler/issues/3041 to be fixed. After this, single pass can simply use `node_modules/google-closure-compiler-java/compiler.jar` with no changes.
2. Delete `third_party/closure-compiler/**` and `build-system/runner/dist/runner.jar`
3. Check in a copy of `ant` to `third_party/ant/`
4. Whenever `gulp dist` or `gulp check-types` are invoked, run a pre-requisite task that generates `runner.jar` by calling `ant clean && ant jar && ant test` and makes sure everything worked
 
**Notes:**
- Whenever a new version of closure compiler is released, a `renovate` PR will automatically test it out by generating a new `runner.jar`, running `gulp check-types`, performing a minified build, and running all our tests.
- If a breaking change appears, we will be alerted by build / test failures, and can make the required changes and move on.
- We'll no longer need to separately maintain `compiler-and-tests.jar` since we are no longer using a custom version of `compiler.jar`. The tests in `build-system/runner` should be sufficient to verify the custom AMP passes.
- The speed up gains for multi pass due to `nailgun` will remain as they are, since we'll continue to re-use a single instance of the dynamically generated `runner.jar`

### PR DESCRIPTION
**Pull request by rsimha**
In #23417, we switched from the custom-built `third_party/closure-compiler/compiler.jar` to the natively installed `node_modules/google-closure-compiler-java/compiler.jar`. With this, the coast is clear to dynamically generate `runner.jar` instead of checking in a pre-built copy.

**PR highlights:**
- ~Adds a copy of [Apache Ant](https://ant.apache.org/) v1.10.6 to `third_party/ant/`~ **Edit:** Moved to #23490
- Removes all unnecessary files from `build-system/runner` including the pre-built `runner.jar`
- Adds `build-system/tasks/generate-runner.js`, which dynamically generates `runner.jar` whenever a `gulp` task needs to invoke closure compiler (`dist`, `check-types`, and `nailgun-start`)

**Notes:**
- The dynamically generated `runner.jar` is written to a subdirectory of `build-system/runner/dist/` in order to allow for the concurrent running of `gulp dist` and `gulp check-types`
- Information about the commit at which `runner.jar` was last generated is written to the directory, so that it is only regenerated when required
- It is no longer necessary to manually update `runner.jar` while upgrading closure compiler
- Changes to `build-system/runner/src/org/ampproject/*.java` are automatically applied when you run `gulp dist` or `gulp check-types`
- New releases of closure compiler will automatically create a Renovate PR whose Travis build will generate and test a new `runner.jar`

Fixes #22452

### CODE REVIEW

**Comment by rsimha** [comment]
@jridgewell and @estherkim have helped verify that this works on older machines, and machines that have never had to run `ant` before. (Thanks!)

**Comment by jridgewell** [comment]
A few questions:

- Does this have to be generated every time?
- Why is a checked-in ant better than a checked in runner?

**Comment by rsimha** [comment]
@jridgewell answers below

> * Does this have to be generated every time?

- There's no good way to tell if there are new (local?) changes in `build-system/runner/` that require  `runner.jar` to be regenerated.
- With this PR, we generate a separate instance of `runner.jar` for each instance of `nailgun-server.jar` which allows you to concurrently run `gulp dist` and `gulp check-types`
- Generating `runner.jar` takes O(milliseconds) time, and does not materially increase the running time of `gulp` tasks that use it

> * Why is a checked-in ant better than a checked in runner?

- With a checked-in `runner.jar`, updating our custom closure code requires a three step process:
    - Edit `.java` file(s) in `build-system/runner`
    - Run `ant jar` to generate a new `runner.jar`
    - Commit, merge
    - Note: This has been problematic in the past due to different `ant` versions on dev workstations (see https://github.com/ampproject/amphtml/pull/22485#pullrequestreview-241548765)
- With a checked-in `ant`, updating custom closure code becomes simpler:
    - Edit `.java` file(s) in `build-system/runner`
    - Test commit by running `gulp check-types` or `gulp dist`, or upload to Travis

**Comment by jridgewell** [comment]
I don't think this is a good change to make.

> - There's no good way to tell if there are new changes (local edits? synced changes?) in build-system/runner/ that require runner.jar to be regenerated
> - Generating runner.jar takes ~1s to run, and does not materially increase the running time of gulp tasks that use it

That's just unnecessary extra work, though. (And it took 4s on my machine). I understand that it's small compared to overall closure run time, so this isn't a blocker.


> - With this PR, we generate / use a separate instance of runner.jar for each instance of nailgun-server.jar which allows you to concurrently run gulp dist and gulp check-types

Running closure makes my machine usable for ~5min. Running both dist and check-types at the same time would make my machine _even more_ unusable.

> - With a checked-in runner.jar, updating custom closure code requires these steps:
> - With a checked-in ant, updating custom closure code becomes simpler:

This is really only run buy You, Erwin, and myself. I think a checked in bash script to automate this would be better.

**Comment by rsimha** [comment]
> This is really only run buy You, Erwin, and myself. I think a checked in bash script to automate this would be better.

Requiring a bash script to be run manually will prevent the real benefit of this PR from being realized: The ability to automatically test new closure compiler versions on Travis as they are released (via Renovate PRs). Today, upgrading closure compiler is a painstaking manual process that happens only when someone sets out to do so. Dynamically generating `runner.jar` will significantly reduce the friction involved.

> Running closure makes my machine usable for ~5min. Running both dist and check-types at the same time would make my machine _even more_ unusable.

This is only a problem on old machines like yours. Running concurrent instances of closure is entirely possible on current macbooks and standard linux workstations, and is a fairly useful workflow.

**Comment by jridgewell** [comment]
> Requiring a bash script to be run manually will prevent the real benefit of this PR from being realized: The ability to automatically test new closure compiler versions on Travis as they are released (via Renovate PRs)

That's also easily done via a gulp task before running?

> This is only a problem on old machines like yours. Running concurrent instances of closure is entirely possible on current macbooks and standard linux workstations, and is a fairly useful workflow.

Man, I wish Apple would make a decent laptop again.

Current gen: https://browser.geekbench.com/v4/cpu/13995682
Mine: https://browser.geekbench.com/v4/cpu/13995433

**Comment by rsimha** [comment]
> That's also easily done via a gulp task before running?

As discussed in person, this isn't possible, since it will result in the need to add a new commit to a Github PR from a Travis build.

**Comment by rsimha** [comment]
@jridgewell I've added commit https://github.com/ampproject/amphtml/pull/23434/commits/44f00ca4fdf3e460b0c9df8022bdee008231d96b  that generates `runner.jar` only in the following cases:

1. `runner.jar` hasn't been generated before
2. The first time a new commit is pulled from `master`
2. There are local changes to `build-system/runner/`

Let me know what you think.

File: build-system/tasks/generate-runner.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    return Promise.reject(reason);
+  } else {
+    if (!isTravisBuild()) {
+      log(
+        'Generated custom closure compiler',
+        cyan(`${runnerJarDir}/runner.jar`)
+      );
+    }
+  }
+  return Promise.resolve();
```
Comment (erwinmombay) [inline build-system/tasks/generate-runner.js]: don't believe you need this since this is an async function

File: third_party/ant/LICENSE [hunk 2]
```diff
@@ -0,0 +1,272 @@
+/*
```
Comment (erwinmombay) [inline third_party/ant/LICENSE]: per malte's old instructions, something like this should be done in its own commit or its own Pull Request

Comment (rsimha) [reply third_party/ant/LICENSE]: This is in a separate commit: https://github.com/ampproject/amphtml/pull/23434/commits/b354627f05636b1b8f81949d9f16ab29df2420c4

Did you mean it should be checked in first and then used in a follow up PR? (We haven't done that with most other third party libraries.)

File: build-system/tasks/generate-runner.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  writeGeneratedAtCommitFile(runnerJarDir);
+  const result = getOutput(generateCmd);
+  if (result.stderr) {
+    log(
+      red('ERROR:'),
+      'Could not generate custom closure compiler',
+      cyan(`${runnerJarDir}/runner.jar`)
+    );
+    console.error(red(result.stdout), red(result.stderr));
+    const reason = new Error('Compiler generation failed');
```
Comment (erwinmombay) [inline build-system/tasks/generate-runner.js]: could we propagate more debugging info here on the why it failed?

Comment (rsimha) [reply build-system/tasks/generate-runner.js]: This is already being done on line 112. The `reason` is merely to let `gulp` know that the task failed.

File: build-system/tasks/generate-runner.js (hunk 1 above)
Comment (rsimha) [reply build-system/tasks/generate-runner.js]: This is a companion to line 115:

https://github.com/ampproject/amphtml/blob/84142ec79c283792ad762af250acd390accf5d97/build-system/tasks/generate-runner.js#L115

I prefer to be explicit and return a Promise in both the success and failure cases.

Comment (jridgewell) [reply build-system/tasks/generate-runner.js]: Async implies a promise, though. This creates a extra object, and forces an extra 3 queue ticks to move to the next step.

File: third_party/ant/LICENSE (hunk 2 above)
Comment (erwinmombay) [reply third_party/ant/LICENSE]: if its in a separate commit, just make sure to not squash when you merge then

File: build-system/tasks/generate-runner.js (hunk 1 above)
Comment (rsimha) [reply build-system/tasks/generate-runner.js]: Removed in https://github.com/ampproject/amphtml/pull/23434/commits/bf08e7b6eefe82a7f63e164ef9076ad47487ab13.

File: third_party/ant/LICENSE (hunk 2 above)
Comment (rsimha) [reply third_party/ant/LICENSE]: That'll create too many commits on `master`. Let me fork it off into its own PR.

**Comment by rsimha** [comment]
@jridgewell @erwinmombay All comments addressed. Third party changes separated out into #23490. PTAL.

File: build-system/git.js [hunk 4]
```diff
@@ -188,6 +188,15 @@ function gitMasterBaseline() {
   return gitMergeBaseLocalMaster();
 }
 
+/**
+ * Returns the diffs for given path based on the given commit
+ * @param {string} path
+ * @param {string} commit
```
Comment (erwinmombay) [inline build-system/git.js]: return {string}

**Comment by erwinmombay** [comment]
just 1 last nit from me. I'll leave it to @jridgewell for final approval since i know he has concerns above

**Comment by rsimha** [comment]
> I still don't like this.

Do you have a better idea to allow renovate to test closure upgrades?

File: build-system/git.js (hunk 4 above)
Comment (rsimha) [reply build-system/git.js]: Fixed.

File: third_party/ant/LICENSE (hunk 2 above)
Comment (rsimha) [reply third_party/ant/LICENSE]: Done.
